Documentation for merge trains

rudymatela opened the pull request:
<!-- NOTE:
Keep in mind that this repository is public. Please avoid posting things like
logs with references to private repositories.
-->

Closes: #169 

This PR documents current behaviour with regards to merge trains:

* [readme.md](https://github.com/channable/hoff/blob/merge-train/documentation/readme.md)

### TODO

- [x] add mention of merge trains on the [readme] file;
- [x] add happy-day diagram without merge trains
- [x] add happy-day diagram with merge trains
- [x] diagram for when build fails
- [x] diagram for when rebase fails
- [x] different icon for cancelling a build :no_entry_sign: 
- [x] describe each case textually
- [x] diagram for delayed build statuses ("wrong" order)
- [x] document when PR is closed (same as when build fails)
- [x] document when new commit arrives in the train (same as when build fails)
- [x] add screenshots
- [x] go around the code adding assorted comments

[readme]: https://github.com/channable/hoff/blob/merge-train/documentation/readme.md

rudymatela commented:
History was quite messy up to this point.  So I squashed my changes now.

rudymatela commented:
@fatho and @ReinierMaas, in this PR I try to document the behaviour of merge trains well so that:

1. in case something strange happens on a sequence of PRs, we can double-check with the docs if it is expected behaviour.  Sometimes what is the correct course of action is unintuitive.  (Builds restarted seemingly out of nothing, but actually because the base branch has failed.);
2. to make it easier to understand the code.

I also slightly improved inline (Haddock) code documentation.

Please let me know your thoughts or if you have any suggestions.

rudymatela commented:
... when reviewing the `readme.md` I recommend looking at the "rich diff".

ReinierMaas commented on readme.md:
```diff
@@ -14,10 +14,16 @@ of software engineering:
+to it. If the build fails, the commits never make it into master, keeping the
+build green at all times.
+
+Merge trains: if a second PR receives a merge comment while a build is still
+happening, Hoff starts a speculative rebase on top of the previous build.  The
+same goes for a third PR.  Assuming all of the three builds pass, the third PR
+is merged after the time of a single build.
```
Maybe we should shortly discuss the possible downsides of speculatively rebasing each pull request. That the choice was made for speculatively rebasing each individual PR which trades CI time for confidence that the individual PRs are buildable points in our history. As opposed to rollups that include multiple PRs in a single merge causing those rollups to act as the guaranteed buildable points in the history. I think that both the simplicity and additional confidence are worth the extra CI cost. The achieved integration speed might also be slightly higher because we don't have to split rollups into multiple separate builds upon failure.

rudymatela replied:
I have added some paraphrased version of your text in the discussion of one of the scenarios -- the one where the "conservative" choice is explained.

ReinierMaas reviewed (changes requested):
Really clear documentation of the advantage of merge trains and their possible failure paths. I am only missing the feedback shortcut that we implemented for orphaned fixups.
Other than that we should add `dia` to the dev tools provided via `nix`.

A slight extension at the start explaining the tradeoffs made with the merge trains is also a good idea.

rudymatela commented:
@ReinierMaas I believe to have addressed your comments in 465bb923c2b5e7696e22947d4cee9cefd55ae6a7, 91d8eac9db7fc51375e1609e69745e110eaf224e and 1724c9c6d0170032028fca897b768022b5485332.  Can you please take a look again?

ReinierMaas commented on default.nix:
```diff
@@ -10,5 +10,6 @@ in
       pkgs.haskellPackages.haskell-language-server
       pkgs.shellcheck
       pkgs.stack
+      pkgs.dia
```
The list looks sorted except for `niv` and `dia`.

ReinierMaas reviewed (commented):
Almost done. Two more nits

rudymatela commented:
> Almost done. Two more nits

:runner: Fixed! :smile:

rudymatela commented:
I will merge now so I can prepare a release and deployment for tomorrow early morning.  (The two nits were fixed :grin:  If you have further comments let me know and I can try to address them in further PRs.)

@OpsBotPrime merge

OpsBotPrime commented:
Pull request approved for merge by @rudymatela, rebasing now.

OpsBotPrime commented:
Rebased as e0cd286121a38e89561e0dfc732ac6de4bb73d9b, waiting for CI …

OpsBotPrime commented:
[CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/23aec422-b387-4684-9884-c391efea9c96?pipeline_id=18544702-af78-41d8-a533-a734ab630211) started.
